### ISSUE TITLE
Better support for abstract classes which does not generate dead code.

### ISSUE THREAD

**Issue by mhevery** [angular/tsickle#505]
## Overview

Currently abstract classes produce a lot of runtime dead code which can not be tree shaken by uglify. The core of the issue is that tsickle converts abstract methods into concrete methods for closure which are than retained by uglify.

Here is a proposal for new way of doing it which does not produce dead code.

Given
```
abstract class Base {
  abstract methodA();
}
```

Tsickle currently generates
```
class Base {
  /**
   * @abstract
   * @return {void}
   */
  methodA() {}
}
 ```

which than downlevels to:
```
var Base=(function(){
  function Base(){}
  Base.prototoype.methodA = function () {};
  return Base;
}());
```
 
Notice that the abstract method now is no longer abstract and it has a method body. This creates an issue because the method is now retained at runtime even though it is never called (it was abstract)
 
## Proposal
Here is a proposal for how the tsickle should generate the code instead. ([playground](https://jscomp-debug.googleplex.com/#input0%3Dclass%2520Base%2520%257B%250A%2520%2520constructor()%2520%257B%250A%2520%2520%2520%2520%252F**%2520%2540type%2520%257BFunction%257D%2520*%252F%250A%2520%2520%2520%2520this.methodA%253B%250A%2520%2520%257D%250A%257D%250A%250Aclass%2520Impl%2520extends%2520Base%2520%257B%250A%2520%2520methodA()%2520%257B%2520%250A%2520%2520%2520%2520console.log('impl')%2520%250A%2520%2520%257D%253B%250A%257D%250A%250A%252F**%2520%2540param%2520%257BBase%257D%2520x%2520*%252F%250Afunction%2520useBase(x)%2520%257B%250A%2520x.methodA()%253B%250A%257D%250A%250AuseBase(new%2520Impl())%253B%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26MISSING_PROPERTIES%3D1%26TRANSPILE%3D1%26CHECK_TYPES%3D1%26COMPUTE_FUNCTION_SIDE_EFFECTS%3D1%26INLINE_VARIABLES%3D1%26SMART_NAME_REMOVAL%3D1%26REMOVE_UNUSED_PROTOTYPE_PROPERTIES%3D1%26REMOVE_UNUSED_PROTOTYPE_PROPERTIES_IN_EXTERNS%3D1%26COLLAPSE_PROPERTIES%3D1%26DEVIRTUALIZE_PROTOTYPE_METHODS%3D1%26VARIABLE_RENAMING%3D1%26PROPERTY_RENAMING%3D1%26CLOSURE_PASS%3D1%26GENERATE_EXPORTS%3D1%26PRETTY_PRINT%3D1))

```
class Base {
  constructor() {
    /** @type {Function} */ this.methodA;
  }
}
```
 
The advantage is that uglifyJS (uglifyjs --compress "pure_getters=true") can now correctly process the output:
1. there is no actual method in the source.
2. The types are satisfied through a field  `/** @type {Function} */ this.methodA;`.
3. uglify can remove `this.methodA` since it is sure that it has no side-effects.

Resulting in:
```
var Base=(function(){
  function Base(){}
  return Base;
}());
```
 
Notice that there are no abstract references, either in prototype or constructor.

**Comment by evmar** [comment]
Note to others: this is different from other tsickle requests because the point is to generate code that simultaneously works in Closure and can be eliminated in uglify.

**Comment by mhevery** [comment]
Little addition after talking with @rkirov about corner cases with the constructor. In some cases the tsickle will have to create a constructor where there was no constructor before. In such a case tsickle could keep the old behavior as in practice this would be very rare.

The condition under which this arises are:
1) There is an abstract subclass which extends existing class. 
2) The subclass itself does not have a constructor declared.
3) The subclass adds additional abstract methods.

**Comment by mhevery** [comment]
After more taking to @rkirov I can to conclusion that abstract methods should be treated in the same way as fields of the same type. 

Hence 
```
abstract class Base {
  abstract methodA()
}
```

Should behave same as 

```
abstract class Base {
  methodA: () => void;
}
```

**Comment by rkirov** [comment]
Hmm, now TypeScript objects at the actual implementation, because it tries to keep methods and fields separate.

```typescript
abstract class Base {
    methodA: () => void; 
}

class Extends extends Base {
    methodA(){ };
}
```
errors with `Class 'Base' defines instance member property 'methodA', but extended class 'Extends' defines it as instance member function.`

I would be nice to avoid rewriting the extending subclass due to info from Base, so this solution might be a no go too.

**Comment by evmar** [comment]
We don't typecheck after codegen, so we shouldn't get any type errors after translating in this manner.

**Comment by mprobst** [comment]
Let me take a look at this.

**Comment by mhevery** [comment]
closed by #506
